### ISSUE TITLE
AssertionError in the printing trace tree

### ISSUE THREAD

**Issue by pierrehvz** [openfisca/openfisca-core#403]
I was writing tests for isf in openfisca and I was using your tool but it is not working for me with theses two variables (  `isf_tot` and `isf_apres_plaf` ) and I don't understand why:

My code is: 

```
from openfisca_france import FranceTaxBenefitSystem
tax_benefit_system = FranceTaxBenefitSystem()


scenario = tax_benefit_system.new_scenario().init_single_entity(
    parent1=dict(salaire_de_base=150000,
                 age = 60,
                 ),
    period=2016,
    foyer_fiscal = dict(
        b1ab = 2000000,
        b1bc = 100000,
        b1be = 100000,
        b1cf = 100000,
        b2nc = 1000
        ),
    )
simulation = scenario.new_simulation(trace=True)
simulation.calculate('isf_apres_plaf', 2016, print_trace=True)
```

and the error message is:

```
assert step is not None 
AssertionError
```

Thank you for your help !! @cbenz

**Comment by cbenz** [comment]
My fault probably :-( I fix it in fireman mode.

**Comment by cbenz** [comment]
Sorry but I copy-paste your script using the latest OpenFisca-Core and France versions and it works.

What's your complete stacktrace and which versions do you use? Hint:

```
pip freeze | grep -i OpenFisca-
```

My result:

```
(openfisca-fix-trace) ~/D/o/openfisca-core-fix-trace (fix-trace) $ pip freeze | grep -i OpenFisca-
OpenFisca-Core==2.2.1
-e git+git@github.com:openfisca/openfisca-france.git@945ad448a627b8e76a600ffaa7235acfc57a77a2#egg=OpenFisca_France
```

**Comment by benjello** [comment]
@pierrehvz actually used this working [branch](https://github.com/pierrehvz/openfisca-france/tree/isf)
And you should use it with the baremes-ipp branch of openfisca-core.

**Comment by pierrehvz** [comment]
Pour information, en tapant:
`pip freeze | grep -i OpenFisca-`

j'obtiens:
```
-e git+https://github.com/openfisca/openfisca-core.git@cce2f72ddfa93616b2320a38ab9f56164f6b9654#egg=OpenFisca_Core
-e git+https://github.com/pierrehvz/openfisca-france.git@0ba495a5d57ee1af4f96c38d196d2bb8aabd877c#egg=OpenFisca_France
OpenFisca-Matplotlib==0.4.dev0
```

**Comment by cbenz** [comment]
I have for myself this error:

```
(openfisca-fix-trace) ~/D/o/openfisca-core-fix-trace (baremes-ipp) $ python test.py 
Traceback (most recent call last):
  File "test.py", line 19, in <module>
    simulation.calculate('isf_apres_plaf', 2016, print_trace=True)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/simulations.py", line 74, in calculate
    return self.compute(column_name, period = period, **parameters).array
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/simulations.py", line 159, in compute
    result = holder.compute(period = period, **parameters)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/holders.py", line 137, in compute
    formula_dated_holder = self.formula.compute(period = period, **parameters)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/formulas.py", line 269, in compute
    dated_holder = dated_formula['formula'].compute(period = output_period, **parameters)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/formulas.py", line 555, in compute
    formula_result = self.base_function(simulation, period)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/base_functions.py", line 60, in requested_period_default_value
    return formula.function(simulation, period, *extra_params)
  File "/home/cbenz/Dev/openfisca/openfisca-france-baremes-ipp/openfisca_france/model/prelevements_obligatoires/isf.py", line 555, in function_20130101_20151231
    tot_impot = simulation.calculate('tot_impot', period)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/simulations.py", line 74, in calculate
    return self.compute(column_name, period = period, **parameters).array
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/simulations.py", line 159, in compute
    result = holder.compute(period = period, **parameters)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/holders.py", line 137, in compute
    formula_dated_holder = self.formula.compute(period = period, **parameters)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/formulas.py", line 555, in compute
    formula_result = self.base_function(simulation, period)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/base_functions.py", line 60, in requested_period_default_value
    return formula.function(simulation, period, *extra_params)
  File "/home/cbenz/Dev/openfisca/openfisca-france-baremes-ipp/openfisca_france/model/prelevements_obligatoires/isf.py", line 436, in function
    irpp = simulation.calculate('irpp', period)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/simulations.py", line 74, in calculate
    return self.compute(column_name, period = period, **parameters).array
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/simulations.py", line 159, in compute
    result = holder.compute(period = period, **parameters)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/holders.py", line 137, in compute
    formula_dated_holder = self.formula.compute(period = period, **parameters)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/formulas.py", line 555, in compute
    formula_result = self.base_function(simulation, period)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/base_functions.py", line 60, in requested_period_default_value
    return formula.function(simulation, period, *extra_params)
  File "/home/cbenz/Dev/openfisca/openfisca-france-baremes-ipp/openfisca_france/model/prelevements_obligatoires/impot_revenu/ir.py", line 1636, in function
    iai = simulation.calculate('iai', period)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/simulations.py", line 74, in calculate
    return self.compute(column_name, period = period, **parameters).array
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/simulations.py", line 159, in compute
    result = holder.compute(period = period, **parameters)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/holders.py", line 137, in compute
    formula_dated_holder = self.formula.compute(period = period, **parameters)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/formulas.py", line 555, in compute
    formula_result = self.base_function(simulation, period)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/base_functions.py", line 60, in requested_period_default_value
    return formula.function(simulation, period, *extra_params)
  File "/home/cbenz/Dev/openfisca/openfisca-france-baremes-ipp/openfisca_france/model/prelevements_obligatoires/impot_revenu/ir.py", line 1599, in function
    teicaa = simulation.calculate('teicaa', period)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/simulations.py", line 74, in calculate
    return self.compute(column_name, period = period, **parameters).array
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/simulations.py", line 159, in compute
    result = holder.compute(period = period, **parameters)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/holders.py", line 137, in compute
    formula_dated_holder = self.formula.compute(period = period, **parameters)
  File "/home/cbenz/Dev/openfisca/openfisca-core-fix-trace/openfisca_core/formulas.py", line 568, in compute
    variable_name = column.name,
openfisca_core.legislations.ParameterNotFound: Legislation parameter "impot_revenu.teicaa" was not found at instant "2016-01-01" by variable "teicaa"
```

which does not seem to fail at the same point.

Again, could you bring your full stack trace? Thanks!

**Comment by pierrehvz** [comment]
My stack trace:
```

λ python "..\..\Python Scripts\simul_isf.py"
C:\Users\p.havez\AppData\Local\Continuum\Anaconda2\lib\site-packages\numpy\core\fromnumeric.py:2786: RuntimeWarning: invalid value encountered in rint
  return round(decimals, out)
isf_apres_plaf@foyers_fiscaux<2016>[0.0]
|---> aide_logement_base_ressources_defaut@familles<2016-04>[0.0]
|     |---> aah@individus<2015-12>[0.0]
|     |     |---> aah_base@individus<2015-12>[0.0]
|     |---> asi_eligibilite@individus<2015-12>[False]
|     |     |---> aspa_eligibilite@individus<2015-12>[False]
|     |     |---> handicap@individus<2015-12>[False]
|     |     |---> retraite_nette@individus<2015-11>[0.0]
|     |     |---> pensions_invalidite@individus<2015-11>[0.0]
|     |     |---> asi_aspa_condition_nationalite@individus<2015-12>[True]
|     |---> asi@familles<2015-12>[0.0]
|     |     |---> asi_eligibilite@individus<2015-12>[False]
|     |     |---> aspa_eligibilite@individus<2015-12>[False]
|     |     |---> maries@familles<2015-12>[False]
|     |     |---> en_couple@familles<2015-12>[False]
|     |     |---> asi_aspa_nb_alloc@familles<2015-12>[0]
|     |     |---> asi_aspa_base_ressources@familles<2015-12>[0.0]
Traceback (most recent call last):
  File "..\..\Python Scripts\simul_isf.py", line 26, in <module>
    simulation.calculate('isf_apres_plaf', 2016, print_trace=True)
  File "c:\users\p.havez\documents\github\openfisca-core\openfisca_core\simulations.py", line 74, in calculate
    return self.compute(column_name, period = period, **parameters).array
  File "c:\users\p.havez\documents\github\openfisca-core\openfisca_core\simulations.py", line 161, in compute
    self.print_trace(variable_name=column_name, period=period, **print_trace_kwargs)
  File "c:\users\p.havez\documents\github\openfisca-core\openfisca_core\simulations.py", line 308, in print_trace
    traverse(variable_name, period, depth=0)
  File "c:\users\p.havez\documents\github\openfisca-core\openfisca_core\simulations.py", line 293, in traverse
    traverse(child_variable_name, child_period, depth + 1)
  File "c:\users\p.havez\documents\github\openfisca-core\openfisca_core\simulations.py", line 293, in traverse
    traverse(child_variable_name, child_period, depth + 1)
  File "c:\users\p.havez\documents\github\openfisca-core\openfisca_core\simulations.py", line 282, in traverse
    assert step is not None
AssertionError
```

**Comment by cbenz** [comment]
(Je repasse en français !)

Le problème ici est qu'une étape de la trace référence une variable à une période qui n'existe pas (ou n'est pas trouvée).

Plutôt que de planter je peux afficher une indication comme quoi il y a un problème bien que ce ne soit pas satisfaisant à 100% car l'erreur ne devrait pas arriver. Il faut arriver à comprendre pourquoi elle se produit et pour cela je dois pouvoir reproduire le problème.

Je vais continuer à chercher sur la base de ta stack trace et du script `simul_isf.py` dans ta branche.

**Comment by cbenz** [comment]
@pierrehvz Where is your script `simul_isf.py`?

I really need to reproduce the problem. By using the script copy-pasted at the beginning of the issue it works.

I have to ask you to give the full details about the context:

- commit the script which produces an error in your branch (tell me which repo and which branch)
- tell me which branch of openfisca-core you are on, the same for openfiscsa-france
- which line do you type in your terminal to run the script

At that point I hope I will be able to reproduce the error and fix it.

**Comment by pierrehvz** [comment]
- Mon script `simul_isf.py` est dans la branche `isf` dans le repo `openfisca-france`
- Concernant `openfisca-core` et `openfisca-france`, je suis sur la branche `baremes-ipp`.
- Je n'utilise pas le terminal, je lance le script directement sur spyder

Merci @cbenz

**Comment by cbenz** [comment]
I'm back on this issue

**Comment by MattiSG** [comment]
@pierrehvz Is this issue still relevant for you?

**Comment by MattiSG** [comment]
No work done on this in the last 5 months, closing. @pierrehvz please let me know if this is still not fixed and should be reopened!  😃

**Comment by cbenz** [comment]
@MattiSG For your information @pierrehvz was a trainee at the IPP and is no more involved in OpenFisca, excepted for personal reasons ;-)

@benjello will probably be the one to ask for this feature to be fixed.

**Comment by benjello** [comment]
Yep we definitely need to fix the print_trace which needs neat parsers IIRC.

**Comment by MattiSG** [comment]
Ok, but the issue seemed to only be present in a specific branch. It could be present in `master` now that `baremes-ipp` has been merged, but this needs investigation.

Also, since the error appears to be linked to the periods, this could be fixed by #452. Let's wait until #452 is fixed to see if this issue is still relevant (or simply wait until someone encounters the problem again :wink: ).

**Comment by michelbl** [comment]
`print_trace` is removed by #452 ...
